File: PXUltima/Source/Format/Color.c
#include "Color.h"

void ConvertColorRGBToYCbCr(const ColorRGB* const colorRGB, ColorYCbCr* const ColorYCbCr)
{
	ColorYCbCr->Y = 0.299 * colorRGB->Red + 0.587 * colorRGB->Green + 0.114 * colorRGB->Blue;
	ColorYCbCr->Cb = -0.1687 * colorRGB->Red - 0.3313 * colorRGB->Green + 0.5 * colorRGB->Blue + 128u;
	ColorYCbCr->Cr = 0.5 * colorRGB->Red - 0.4187 * colorRGB->Green - 0.0813 * colorRGB->Blue + 128u;
}

void ConvertColorYCbCrToRGB(const ColorYCbCr* const colorYCbCr, ColorRGB* const colorRGB)
{
	colorRGB->Red = colorYCbCr->Y + 1.402 * (colorYCbCr->Cr - 128u);
	colorRGB->Green = colorYCbCr->Y - 0.34414 * (colorYCbCr->Cb - 128u) - 0.71414 * (colorYCbCr->Cr - 128u);
	colorRGB->Blue = colorYCbCr->Y + 1.772 * (colorYCbCr->Cb - 128u);
}

float ConvertColorRGBToY(const float red, const float green, const float blue)
{
	return +0.299f * red + 0.587f * green + 0.114f * blue;
}

float ConvertColorRGBToCb(const float red, const float green, const float blue)
{
	return -0.16874f * red - 0.33126f * green + 0.5f * blue;
} // ITU: -0.168736f * r -0.331264f * g +0.5f      * b

float ConvertColorRGBToCr(const float red, const float green, const float blue)
{
	return +0.5f * red - 0.41869f * green - 0.08131f * blue;
} // ITU: +0.5f      * r -0.418688f * g -0.081312f * b

File: Code/PXUltimaCS/Source/Media/Color.cs
namespace PX
{
    public enum PXColorFormat
    {
        PXColorFormatInvalid,

        PXColorFormatAlphaBinary,

        // Single channel
        PXColorFormatR8,
        PXColorFormatG8,
        PXColorFormatB8,
        PXColorFormatA8,

        PXColorFormatR16,
        PXColorFormatG16,
        PXColorFormatB16,
        PXColorFormatA16,

        PXColorFormatR32,
        PXColorFormatG32,
        PXColorFormatB32,
        PXColorFormatA32,

        PXColorFormatRF,
        PXColorFormatGF,
        PXColorFormatBF,
        PXColorFormatAF,

        PXColorFormatRD,
        PXColorFormatGD,
        PXColorFormatBD,
        PXColorFormatAD,



        // RGB
        PXColorFormatRGBI8,
        PXColorFormatRGBI16,
        PXColorFormatRGBI32,
        PXColorFormatRGBF,
        PXColorFormatRGBD,

        // BGR
        PXColorFormatBGRI8,
        PXColorFormatBGRI16,
        PXColorFormatBGRI32,
        PXColorFormatBGRF,
        PXColorFormatBGRD,

        // RGBA
        PXColorFormatRGBAI8,
        PXColorFormatRGBAI16,
        PXColorFormatRGBAI32,
        PXColorFormatRGBAF,
        PXColorFormatRGBAD,

        // BGRA
        PXColorFormatBGRAI8,
        PXColorFormatBGRAI16,
        PXColorFormatBGRAI32,
        PXColorFormatBGRAF,
        PXColorFormatBGRAD,




        PXColorFormatYCbCrI8,







        PXColorFormatARGBI8,
        PXColorFormatARGBF16,
        PXColorFormatARGBF32,

        PXColorFormatXRGBI8,

        PXColorFormatRGBI5,





        PXColorFormatLuminance,
        PXColorFormatLuminanceAlpha,
        PXColorFormatColorIndex,
        PXColorFormatStencilIndex,
        PXColorFormatDepthComponent
    }

    public struct PXColorRGBI8
    {
        byte Red;
        byte Green;
        byte Blue;
    }

    public struct PXColorRGBI16
    {
        ushort Red;
        ushort Green;
        ushort Blue;
    }

    public struct PXColorRGBI32
    {
        uint Red;
        uint Green;
        uint Blue;
    }
    public struct PXColorRGBF
    {
        float Red;
        float Green;
        float Blue;
    }

    public struct PXColorRGBAF
    {
        public float Red;
        public float Green;
        public float Blue;
        public float Alpha;
    }

    public struct PXColorRGBD
    {
        public double Red;
        public double Green;
        public double Blue;
    }

    public struct PXColorYCbCrI8
    {
        byte Y;
        byte Cb;
        byte Cr;
    }

    //---[ Public-Functions ]----------------------------------------------

    public class Color
    {
        public static void RGBToYCbCr(ref PXColorRGBI8 colorRGB, ref PXColorYCbCrI8 ColorYCbCr)
        {
        
        }
        public static void YCbCrToRGB(ref PXColorYCbCrI8 colorYCbCr, ref PXColorRGBI8 colorRGB)
        {
            
        }

        public static float RGBToY(float red, float green, float blue)
        {
            return 0;
        }
        public static float RGBToCb(float red, float green, float blue)
        {
            return 0;
        }
        public static float RGBToCr(float red, float green, float blue)
        {
            return 0;
        }
    }
}
